### PR TITLE
fix and test for issue #2811

### PR DESCRIPTION
**Pull request by xwang777**
- [ ] In dask.dataframe.sample(...), check empty df and return it if so.
- [ ] In dask/dataframe/tests/test_dataframe.py, added test_sample_from_query. It fails before the fix and passes after the fix.

### CODE REVIEW

**Comment by xwang777** [comment]
In test_dataframe.py, I'm not sure how to do the dask-way of assert, use I uses python assert instead.

File: dask/dataframe/tests/test_dataframe.py [hunk 1]
```diff
@@ -2727,3 +2727,17 @@ def test_better_errors_object_reductions():
...
+        ret['key'] = key
+        return ret
+    ddf = dd.from_delayed([ dask.delayed(_mk_df)(k) for k in range(100) ])
+    ddf = ddf.repartition(npartitions=4)
+    assert ddf.npartitions == 4
+
+    # most of the partitions don't have key > 90
+    ddf1 = ddf.query("key > 90").sample(frac=0.1)
+    df = ddf1.compute()
+    assert np.all(df['key'] > 90)
```
Comment (jcrist) [inline dask/dataframe/tests/test_dataframe.py]: A simpler test (with a title more fitting of the issue):

```python
def test_sample_empty_partitions():
    @dask.delayed
    def make_df(n):
        return pd.DataFrame(np.zeros((n, 4)), columns=list('abcd'))
    ddf = dd.from_delayed([make_df(0), make_df(100), make_df(0)])
    ddf2 = ddf.sample(frac=0.2)
    # smoke test sample on empty partitions
    res = ddf2.compute()
    assert res.dtypes.equals(ddf2.dtypes)
```

Comment (TomAugspurger) [reply dask/dataframe/tests/test_dataframe.py]: Can we simplify further?

```python
In [16]: df = pd.DataFrame({"A": [True, True, False, False]})

In [17]: ddf = dd.from_pandas(df, 2)

In [18]: ddf[ddf.A].sample(frac=.5).compute()
```

I haven't tested that, but I think it hits the same thing?

Comment (jcrist) [reply dask/dataframe/tests/test_dataframe.py]: I'd say that this is less clear that you have empty partitions, but sure, that tests the same thing.

Comment (xwang777) [reply dask/dataframe/tests/test_dataframe.py]: I think jcrist's simplified test is good. Do I need to anything for this to be merged? Thanks.

Comment (jcrist) [reply dask/dataframe/tests/test_dataframe.py]: > Do I need to anything for this to be merged?

Can you update the test as I described above?

Comment (mrocklin) [reply dask/dataframe/tests/test_dataframe.py]: @xwang777 are you able to update this PR with the change suggested above?  If so then we can go ahead and merge this.  (thanks for your work!)

**Comment by xwang777** [comment]
I committed the update. Sorry that it got delayed because of other things.
Thanks.

**Comment by mrocklin** [comment]
Looks good to me.  We might also report this upstream to Pandas

```python
In [1]: import pandas as pd

In [2]: df = pd.DataFrame({'x': [1, 2, 3]})

In [3]: df.iloc[:0].sample(frac=0.5)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-ba2ef6b4fea3> in <module>()
----> 1 df.iloc[:0].sample(frac=0.5)

/home/mrocklin/Software/anaconda/lib/python3.6/site-packages/pandas/core/generic.py in sample(self, n, frac, replace, weights, random_state, axis)
   3439                              "provide positive value.")
   3440 
-> 3441         locs = rs.choice(axis_length, size=n, replace=replace, p=weights)
   3442         return self.take(locs, axis=axis, is_copy=False)
   3443 

mtrand.pyx in mtrand.RandomState.choice (numpy/random/mtrand/mtrand.c:17104)()

ValueError: a must be greater than 0
```

**Comment by mrocklin** [comment]
It looks like you have a small flake8 error

**Comment by xwang777** [comment]
What's flake8 error? I tried the test and it ran OK.

**Comment by TomAugspurger** [comment]
flake8 is a style checker. You can see the error at https://travis-ci.org/dask/dask/jobs/295497416#L4892, or install `flake8` and run it locally.

```
dask/dataframe/tests/test_dataframe.py:2767:1: E302 expected 2 blank lines, found 1
```

You'll want to add another blank line there.

**Comment by xwang777** [comment]
I added the blank line and it passed flake8. Good to know. Thanks.

**Comment by mrocklin** [comment]
Thanks @xwang777 !  Merging

**Pull request merged**